### PR TITLE
Rails 6

### LINKED ISSUE #289: Update dependencies
More and more security vulnerabilities are popping up notified by GitHub. It feels like in general, we're getting out of date on our dependencies. We should put our efforts on doing a major bum up so that it's easy and painless to apply single upgrades when needed.

This deserves some investigation so that we can come up with a strategy to do it as I assume most of the gems will need decent upgrades.

### PR DESCRIPTION
**Pull request by markets**
# Rails 6 🎉 

Extras:
- :gem: updates :arrow_up: (closes #289, closes #567, closes #566)
- Sprockets v4
- Removed `coffee-rails`
- jquery2 (faster and smaller)
- DRY staging env 
- Clean up .gitignore file
- Removed deprecated rollbar:deploy script
- Introduce `.rspec` file

Notes:
- Arguments changes in tests has been fixed automatically by the `Rails/HttpPositionalArguments` cop
- Coffee to JS, by https://decaffeinate-project.org + minor tweaks/refactor (loadTags function to DRY code)
- Fixed all potential security vulnerabilities in deps

### CODE REVIEW

**Comment by markets** [comment]
Current status:
- App boots fine
- Most of parts I tested/played work well
- Tests still not fixed

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -58,11 +58,11 @@ def update
   private
 
   def search_and_load_members(members_scope, default_search_params)
-    @search = members_scope.ransack(default_search_params.merge(params.fetch(:q, {})))
+    @search = members_scope.ransack(default_search_params.merge(params.to_unsafe_h.fetch(:q, {})))
 
     result = @search.result
-    orders = result.orders.map { |order| order.direction == :asc ? "#{order.to_sql} NULLS FIRST" : "#{order.to_sql} NULLS LAST" }
-    result = result.except(:order).order(orders.join(", ")) if orders.count > 0
+    # orders = result.orders.map { |order| order.direction == :asc ? "#{order.to_sql} NULLS FIRST" : "#{order.to_sql} NULLS LAST" }
```
Comment (markets) [inline app/controllers/users_controller.rb]: :warning: this is actually crashing :boom:

**Comment by sauloperez** [comment]
@markets wouldn't we a bit less risky to do a major revamp of all our gems in our current version and then going through a more progressive migration path, from 4.2 to 5, then to 5.2 and finally to 6? again, I'm just aiming to minimize anything that requires our attention.

**Comment by markets** [comment]
Hi @sauloperez 👋 

I don't think so, this way we need to test the app only 1 time. Locally (before the fist commit I mean), I already made the transition from 4.2 -> 5.0 -> 5.2 -> 6.0. I know this is really big branch 😺, but I prefer 1 big branch and test the whole thing 1 time vs several Rails upgrades and tests.

On the other hand, during the latest 4/5 months, I worked on this topic in my company (we upgraded 5 BIG apps from 4.2 -> 6.0), so more or less I have all the "tricks" in my mind.

I'll finalize this after the Elastic ✂️, so I don't need to migrate the Elastic gems.

**Comment by sauloperez** [comment]
Ok, @markets far enough. I give you all my trust :joy: . By the way, those migration skills you got might come in handy in another project you might know about :smirk:

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (markets) [reply app/controllers/users_controller.rb]: Fixed!

**Comment by markets** [comment]
**Update:** only specs still pending to be reviewed. I just made a first attempt, and I got a lot of errors from ElasticSearch:

> Elasticsearch::Transport::Transport::Errors::BadRequest: [400] {"error":{"root_cause":[{"type":"invalid_type_name_exception","reason":"Document mapping type name can't start with '_', found: [_doc]"}],"type":"invalid_type_name_exception","reason":"Document mapping type name can't start with '_', found: [_doc]"},"status":400}

 But as pointed above, I'll prefer to avoid this migration, so I'll wait until the ElasticSearch ✂️.

File: .travis.yml [hunk 2]
```diff
@@ -11,10 +11,9 @@ env:
   global:
     - DATABASE_URL=postgres://postgres@localhost/timeoverflow_test
     - CC_TEST_REPORTER_ID=025bc15a0fa9afa52d86ee24fea845cf1d363f48a466bcf2cef8ab80c29acb28
-before_install:
-  - gem install bundler:1.17.3
 before_script:
-  - bundle exec rake db:setup
+  - psql -c 'create database timeoverflow_test;' -U postgres
```
Comment (sauloperez) [inline .travis.yml:15]: why not use `bundle exec db:create RAILS_ENV=test`? I find it a bit more intention revealing and a bit more future-proof.

File: config/environments/production.rb [hunk 3]
```diff
@@ -14,33 +14,40 @@
...
-  # NGINX, varnish or squid.
-  # config.action_dispatch.rack_cache = true
+  # Ensures that a master key has been made available in either ENV["RAILS_MASTER_KEY"]
+  # or in config/master.key. This key is used to decrypt credentials (and other encrypted files).
+  # config.require_master_key = true
 
   # Disable serving static files from the `/public` folder by default since
   # Apache or NGINX already handles this.
-  config.serve_static_files = true
+  config.public_file_server.enabled = true
```
Comment (sauloperez) [inline config/environments/production.rb:23]: wait. We have this enabled now!? why do we have Nginx then? :see_no_evil: we'll need to fix it but not now.

File: config/environments/staging.rb [hunk 4]
```diff
@@ -1,97 +1,2 @@
...
-  config.active_support.deprecation = :notify
-
-  # Use default logging formatter so that PID and timestamp are not suppressed.
-  config.log_formatter = ::Logger::Formatter.new
-
-  # Do not dump schema after migrations.
-  config.active_record.dump_schema_after_migration = false
-end
+# Use same settings we use for production
+require_relative "production"
```
Comment (sauloperez) [inline config/environments/staging.rb:2]: I totally agree but we need to check all the ENV vars that may differ, especially the SMTP ones used at line 75 above.

**Review by sauloperez — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 2 above)
Comment (markets) [reply .travis.yml:15]: agree, not really sold on this one :) still fighting with Travis... 

About Travis @sauloperez, I got some problems yesterday night, the travis-ci.org is shutting down (in some weeks) and moving to all projects to travis-ci.com (such a bad move Travis 😠), so now all builds have a lot delay and this make the CI feedback a pain.

I think "the good solution" is to move off from Travis, to CircleCI or GitHub-Actions. But for now, I'll migrate to travis-ci.com temporally. I tried it yesterday night (all tests are passing for me locally) but I can't, due to Organization permissions. Could you please help me with that? :))

File: config/environments/production.rb (hunk 3 above)
Comment (markets) [reply config/environments/production.rb:23]: 100% agree, this should be handled by Nginx, but having this fallback doesn't seem a bad idea probably

File: config/environments/staging.rb (hunk 4 above)
Comment (markets) [reply config/environments/staging.rb:2]: Yep, I think we are already fine with that: https://github.com/coopdevs/timeoverflow/blob/c7e87483e93aaf3d69a1ae6bfa167606127c3711/config/environments/production.rb#L75-L88

**Comment by markets** [comment]
@sauloperez I tried also with `db:create` as you suggested (also again with `db:setup` as is supposed to work), but no luck: 
- https://travis-ci.org/github/coopdevs/timeoverflow/builds/750085194
- https://travis-ci.org/github/coopdevs/timeoverflow/builds/750095739

I need to investigate further, but would be amazing if some admin migrate to travis-ci.com (https://docs.travis-ci.com/user/migrate/open-source-repository-migration/#migrating-a-repository) to avoid those long delays (more than 1 hour).

**Comment by sauloperez** [comment]
Ok, after some back and forth and some confusion I successfully migrated it to https://travis-ci.com/ @markets .

**Comment by markets** [comment]
thank you so much @sauloperez!! we are so close 💪🏼

**Comment by sauloperez** [comment]
I can feel it

**Comment by markets** [comment]
Yay! Build is green ❇️  Ready for review ✔️ 

Not sure why 🤷‍♂️  running `bundle exec rails db:setup` in Travis didn't worked well (https://travis-ci.com/github/coopdevs/timeoverflow/builds/209451670#L516), so I finally created the database and imported the schema using the psql client:

```
- psql -c 'create database timeoverflow_test;' -U postgres
- psql -U postgres -q -d timeoverflow_test -f db/structure.sql
```

At least, it's faster 😄

**Comment by sauloperez** [comment]
Cool! that's not a big deal. Someone will fix it if that ever breaks but I suspect we'll move away from Travis before that :joy:

**Review by sauloperez — APPROVED** [review]
Awesome work @markets ! :metal:

**Comment by sauloperez** [comment]
I tried to deploy it to staging but we need more changes to provisioning. Capistrano raises the following:

```
bundle stderr: /home/timeoverflow/.rbenv/versions/2.6.3/lib/ruby/2.6.0/rubygems.rb:283:in `find_spec_for_exe': Could not find 'bundler' (2.1.4) required by your /var/www/timeoverflow/releases/20201218080546/Gemfile.lock. (Gem::GemNotFoundException)
To update to the latest version installed on your system, run `bundle update --bundler`.
To install the missing version, run `gem install bundler:2.1.4`                                                                 
        from /home/timeoverflow/.rbenv/versions/2.6.3/lib/ruby/2.6.0/rubygems.rb:302:in `activate_bin_path'
        from /home/timeoverflow/.rbenv/versions/2.6.3/bin/bundle:23:in `<main>' 
```

**Comment by sauloperez** [comment]
Now we got passed bundler but there's another failure.

```
05:32 deploy:assets:precompile                                                                                                                                                                                                                                  
      01 $HOME/.rbenv/bin/rbenv exec bundle exec rake assets:precompile                                                                                                                                                                                         
      01 rake aborted!                                                                                                                                                                                                                                          
      01 Uglifier::Error: Unexpected token: operator (>). To use ES6 syntax, harmony mode must be enabled with Uglifier.new(:harmony => true).                       
      01 --                                                                                                                                                                                                                                                     
      01  101     ajax: {                                                                                                                                                                                                                                       
      01  102       url: '/tags.json',                                                                                                                                                                                                                          
      01  103       data: function(params) {                                                                                                                                                                                                                    
      01  104         return { term: params.term };                                                                                                                                                                                                             
      01  105       },
      01  106       processResults: function(data, params) {
      01  107         // parse the data into the format expected by Select2
      01  108         return {
      01   =>           results: $.map(data, item => ({
      01  110             id: item,
      01  111             text: item
      01  112           }))
      01  113         };
      01  114       }
      01  115     }
      01  116   });
      01  117 });
```

you should be able to precompile assets locally and debug it @markets. We should also give you permission to deploy to speed things up.

File: app/models/device_token.rb [hunk 5]
```diff
@@ -1,7 +1,6 @@
-class DeviceToken < ActiveRecord::Base
+class DeviceToken < ApplicationRecord
   belongs_to :user
 
-  validates :user_id, presence: true
```
[the remark this answers is not in the document]
Comment (markets) [reply app/models/device_token.rb:4]: That's new Rails default behaviour with `belongs_to`

File: app/models/post.rb [hunk 6]
```diff
@@ -44,8 +44,6 @@ class Post < ActiveRecord::Base
     where(created_at: (1.week.ago.beginning_of_day...DateTime.now.end_of_day))
   }
 
-  validates :user, presence: true
-  validates :category, presence: true
```
[the remark this answers is not in the document]
Comment (markets) [reply app/models/post.rb:48]: yep, exactly (https://github.com/coopdevs/timeoverflow/pull/516#discussion_r545735183)

**Comment by markets** [comment]
Cool @sauloperez! I'll fix that one about js compilation, thanks!

**Comment by markets** [comment]
Fixed by https://github.com/coopdevs/timeoverflow/pull/516/commits/88b898f968a4791ce4761a435e4796211645a803 ✔️

**Comment by sseerrggii** [comment]
Thanks @markets @sauloperez for these great efforts :clap:  :clap:  :clap:

**Comment by sauloperez** [comment]
Wow! all tests passing! :heavy_check_mark: :checkered_flag:

**Comment by sauloperez** [comment]
All yours for testing @sseerrggii . I just deployed it to staging.

**Comment by sseerrggii** [comment]
:+1:

**Comment by sseerrggii** [comment]
Tested!! Its seems to work perfect, maybe it's my imagination but the app flies :small_airplane:  

I needed edit `config/environments/development.rb`  and add  `config.hosts = "timeoverflow.local"` I'm not sure if I should commit that? 

I noticed one problem, but I'm not sure if its Rails 6 related, I think it happened to me sometimes in the past :thinking:

**Comment by sseerrggii** [comment]
Si! Perfecte! Thanks @markets for me it's green @sauloperez 

:heavy_check_mark:

**Comment by markets** [comment]
👍🏼  @sseerrggii you're right! good catch! I just pushed a couple of fixes:

- one for the "All transfers" page (added a test case too)
- proper settings for `config.hosts`

Now all green again, even code coverage and coverage diff ✅

**Comment by sseerrggii** [comment]
@markets @sauloperez I think it's ready to merge 

Who should do the merge?

**Comment by markets** [comment]
🆒 😎  thanks! Rails 6 is in da house 🎉

**Comment by sauloperez** [comment]
Already live in production :sunglasses: although I took us v3.1.0 and v3.1.1 to get it right.

**Pull request merged**